### DISCUSSION TITLE
MP3Player events?

### DISCUSSION THREAD

**Discussion by jsprenkle** [arendst/Tasmota#20917] [Q&A]
Good afternoon,

I've hooked up a [DF Player](https://tasmota.github.io/docs/DFR0299/#wiring) (an MP3 decoder / player) and I have it working. The board provides serial communications in both directions, but I don't see an option to configure Tasmota to accept them. I assume it was never developed?

If I was interested in implementing it is anyone willing to advise me how to find the source and protocol for contributions?
I'm a professional programmer by trade with a lot of hardware experience.

![IMG_20240310_151048956](https://github.com/arendst/Tasmota/assets/427289/bf5fb365-7c0d-4547-822e-753db7a2d96e)

**Comment by sfromis** [comment]
Here's the driver source code:
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_14_mp3.ino
(only using TX, not RX)

**Comment by jsprenkle** [comment]
Thanks, I thought that was the case from the discussions I found.

**Comment by Staars** [comment]
There is an old project from Tasmota-plagrounds:
https://github.com/arendst/Tasmota-Playground/blob/development/experimental/README_MP3.md

Driver code in the same folder as WIP.

There was no ESP32-Tasmota back then.

I remember the most annoying thing was, that there were quite a few of different MP3 player models with slightly different firmware/API's.

**Reply by jsprenkle** [reply]
Gold mine of information there! Thank you.

Not have a reliable explanation of how a device works would be a bit of a project killer!

I connected the MP3 player serial transmit line to a GPIO pin. Then set up a Serial Bridge device to see what was being sent back. So far it doesn't work. This weekend I may have time to put an oscilloscope on the serial line to determine if nothing was sent or I had the Tasmota setup wrong.

Thanks again :)

**Reply by sfromis** [reply]
For a serial setup, you should generally have both TX and RX configured (unless using a driver like the MP3 one only using one of them). And make sure to have the right baud rate set.

**Reply by jsprenkle** [reply]
Thanks for the clarification on the Serial port.

I read the docs and used the following as setup:
SBaudrate 9600
SSerialSend5 1
![image](https://github.com/arendst/Tasmota/assets/427289/b926baf7-29f0-4051-a8e3-4620ca6b09d5)

The MP3 player docs I read said it used 9600 baud, but that might be wrong...

**Reply by sfromis** [reply]
When I do a setup with `SerBr Tx/Rx`, configured to match, I start to see input appearing as `SSerialReceived` messages. Haven't tried it along with another usage of serial.

**Reply by jsprenkle** [reply]
Sorry, I'm not sure I follow.

Did you mean you believe tasmota can only support one software serial device at a time?
(i.e. MP3 player is using the single available software serial device)

I get no notifications of serial received data, though that might also be a hardware or wiring issue.

**Reply by sfromis** [reply]
No, I'm not trying to express a "belief", either way, merely posted how I've seen the serial bridge work.

In general, Tasmota is able to handle more than one serial at a time. Note that I did not say "more than one software serial", which would be clear if there was an index dropdown when configuring `SerBr` pins. Therefore, I'd recommend sticking with the better hardware serial for one, and then adding software serial "on the side".

**Reply by jsprenkle** [reply]
Thanks. I'm not trying to be snarky. I'm not familiar with the code base so things that might be obvious to others aren't to me.
